sdg/components.py

```diff
@@ -91,19 +96,171 @@ def fetch_secret(secret_name, keys):
+
+    def exec_cmd(cmd, **kwarg):
+        if "stdout" in kwarg or "stderr" in kwarg:
+            # If stdout or stderr is explicitly set, remove capture_output to avoid conflicts
+            res = subprocess.run(cmd, text=True, **kwarg)
+        else:
+            # Default behavior (captures output)
+            res = subprocess.run(cmd, text=True, capture_output=True, **kwarg)
+        if res.stdout:
+            print("STDOUT:", res.stdout)
+        if res.stderr:
+            print("STDERR:", res.stderr)
+        if res.returncode != 0:
+            raise RuntimeError(f"CMD {cmd} failed with error code: {res.returncode}")
+        print(f"Command {cmd} succeeded.")
+
+    def is_ssh(uri: str) -> bool:
+        """Checks if a given Git URI is an SSH-based URL."""
+        ssh_patterns = [
+            r"^git@[\w.-]+:.+",
+            r"^ssh://.+",
+        ]
+        return any(re.match(pattern, uri) for pattern in ssh_patterns)
+
+    def get_git_host(repo_url):
+        """Extracts the Git host (e.g., github.com, gitlab.com) from an SSH repository URL."""
+        match = re.match(r"git@([\w.-]+):", repo_url)
+        if match:
+            return match.group(1)  # Extracted host
+        raise ValueError(f"Invalid SSH repository URL: {repo_url}")
+
+    if not taxonomy_repo_secret:
+        username = os.getenv("GIT_USERNAME")
+        token = os.getenv("GIT_TOKEN")
+        ssh_key = os.getenv("GIT_SSH_KEY")
+    else:
+        print("SDG Repo secret specified, fetching...")
+        username, token, ssh_key = fetch_secret(
+            taxonomy_repo_secret, ["GIT_USERNAME", "GIT_TOKEN", "GIT_SSH_KEY"]
+        )
+        print("SDG Repo secret data retrieved.")
+
+    # Whether not provided via env or secret
+    # Assume the repo is public
+    if not username or ssh_key:
+        print("No credentials provided for taxonomy repo, assuming public repo...")
+
+    ca_cert_path = os.getenv("SDG_CA_CERT_PATH")
+    env = os.environ.copy()
+
+    # Todo(HumairAK): should be platform cert
+    # Set custom CA certificate if provided
+    if ca_cert_path:
+        print(f"SSL CA Info detected. ca cert path: {ca_cert_path}")
+        env["GIT_SSL_CAINFO"] = ca_cert_path
+        # Consume this in the process execution environment
+        # This is required for read_taxonomy calls which
+        # Perform nested calls to git clones.
+        os.environ["GIT_SSL_CAINFO"] = ca_cert_path
+
+    # Username/PAT takes precedence over ssh
+    if username and token:
+        print("Configuring Git Credentials...")
+        # Parse the domain from the repository URL
+        parsed_url = urllib.parse.urlparse(repo_url)
+        git_server = parsed_url.netloc
+
+        # Set up Git credential helper
+        exec_cmd(["git", "config", "--global", "credential.helper", "store"], env=env)
+
+        # Save credentials dynamically for any Git server
+        git_credentials_path = "/data/.git-credentials"
```

Comment:
Could you store these using tempfile.NamedTemporaryFile? Not sure if /data is to a PVC but we likely don't want the Git credentials permanently stored there and using a context manager such as with tempfile.NamedTemporaryFile allows it to clean up even if there is an exception below.

Reply:
ah yeah good catch, I did this for the ssh creds but forgot to do it here, will fix this